keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,423 @@
+
+For class `eTP:Cluster` services: Kube-proxy currently returns an answer following its
+`healthz` state. For `eTP:Local` services: Kube-proxy only reports if the service
+for which the load balancer was created for, has a `Ready` endpoint running on
+the node.
+
+This KEP proposes three changes:
+
+1. That Kube-proxy includes its `healthz` state AND its current answer w.r.t the
+  local endpoints, when it answers to the HC for `eTP:Local` services.
+  Kube-proxy is currently defined as "unhealthy" when `2 * syncPeriod` passes in
+  which it knows that it needs to update the data plane (iptables/ipvs), but has
+  not actually done so.
+
+2. That Kube-proxy provides a mechanism for load balancers to do connection
+  draining for terminating Nodes. This is to be done by Kube-proxy by inspecting
+  a field on the Node object which indicates that the "node is
+  terminating/deleting" and when seen, starts failing the HC. When discussing
+  this scenario, the primary case this applied to was: downscaling by the
+  cluster autoscaler (CA). Given that the CA taints the Node which is to
+  be downscaled and deleted, with the taint: `ToBeDeletedByClusterAutoscaler` it
+  would seem most appropriate to use that here. It is unfortunate to have this
+  taint spread around the code base, but for now: no better indicator has been
+  thought of.
+
+3. This KEP does not attempt to align cloud providers for health checking
+   `eTP:Cluster` services. It recognizes cloud providers have valid reasons for
+   doing this differently depending on their implementations. However, it does
+   want to recommend ways for cloud providers to do health checking in a better
+   way on Kubernetes clusters. The KEP hence proposes that a document be added
+   to https://kubernetes.io/docs/concepts/ which can act as a formal guide and
+   be utilized as knowledge sharing with cloud providers.
+
+## Motivation
+
+The motivation for each change is:
+
+1. Not including the `healthz` state can cause Kube-proxy to indicate to a load
+  balancer that it should send traffic to a Node simply because the endpoint is
+  scheduled there, even though Kube-proxy might not be healthy and successfully
+  managed to write the rules required for actually being able to forward traffic
+  to the endpoint.
+
+2. Nodes used as intermediate nexthop for `eTP:Cluster` services would allow all
+  connections passing through the node while it is being terminated, to
+  gracefully shutdown. For `eTP:Local` workloads running on terminating nodes:
+  they get drained simultaneously with Kube-proxy today. This means that there's
+  no guarantee that connection draining will be enabled for them currently. That
+  might be the case if Kube-proxy is terminated before the endpoint, but there's
+  no guarantee of that. If Kube-proxy reacts to the
+  `ToBeDeletedByClusterAutoscaler` taint, then Kube-proxy can start prematurely
+  failing the HC and enable connection draining for these workloads.
```

Comment:
Is there any special mechanism for "enabling connection draining"? Who is handling this? Does this automatically come from failing health checks?

Reply:
> Does this automatically come from failing health checks?

That's it. Load balancers which support connection draining will start draining when the HC fails.

Reply:
To be clear this is *a* form of connection draining - it's providing enough information so that this mode can be engaged.  It's a tool in the toolbox, not a complete solution.